### DISCUSSION TITLE
Option to provide custom Typescript mapping

### DISCUSSION THREAD

**Discussion by ValentinH** [graphql-editor/graphql-zeus#269] [General]
We are using `graphql-zeus` to interact with a Hasura GraphQL server and some common types from Hasura end up being typed as `any` or `unknown`. For example, `uuid`, `timestampz` or `numeric`.

Ideally, I would like to be able to extend the `typeScriptMap` used [here](https://github.com/graphql-editor/graphql-zeus/blob/master/src/TreeToTS/templates/returnedModelTypes.ts#L7), [here](https://github.com/graphql-editor/graphql-zeus/blob/master/src/TreeToTS/templates/resolveValueTypes.ts#L12) and [here](https://github.com/graphql-editor/graphql-zeus/blob/master/src/TreeToTS/templates/returnedTypes.ts#L6).

One approach I have in mind would be to provide a new option `--tsMappingFile` that would read a JSON file like:
```json
{
  "uuid": "string",
  "timestampz": "string",
  "numeric": "string"
}
```
this custom mapping would be added to the default `typeScriptMap` I mentioned above.

If you agree that this could be a nice addition, I'm ready to do a pull request for this (I would probably need a bit of guidance though).

**Comment by ValentinH** [comment]
I explored a bit the codebase and slightly changed the way to support these custom types.
 
Instead of overriding the `typescriptMap`, I created a `customScalarsMap` (which should ideally be coming from a JSON config to make this configurable) and used it like this:
https://github.com/elba-security/graphql-zeus/commit/1cb5233b1c2bc19f9337870b94c1803997051b3c

**Comment by aexol** [comment]
It will be handled soon with functions:
```ts
const result = await chain(HOST)({},{
  URL: (u) => u as string,
  JSON: (u) => u as { settings: string }
})
```

So the scalar type will be return type of the function you provide to scalar resolvers dictionary

**Comment by ValentinH** [comment]
This means that this would be done at runtime, wouldn't it?
The approach I suggest is done at generation time though.

**Comment by aexol** [comment]
Yes but only once per each Scalar definition.

**Reply by ValentinH** [reply]
Will it support TS?

**Comment by CryptoGraffe** [comment]
I have need for a similar thing. I'm using the `citext` extension for postgres, which leads hasura describing the column as a custom type `citext` and so zeus generates a custom type called `citext` but it maps in typescript to type `unknown` ... I just want it to map to TS `string` instead of unknown. Is there any other workaround for this?


```
['citext']: unknown;
```
needs to become

```
['citext']: string;
```

my IDE is unhappy with me

**Comment by aexol** [comment]
Check `graphql-zeus@5.0.3`
https://zeus.graphqleditor.com/page/graphql/scalars.html
How to provide decoders and encoders. This is still fresh and experimental but works and its type safe and tested :)

**Reply by ValentinH** [reply]
This looks nice but I think it would be even nicer to be able to have this done at generation time. Indeed, when having a lot of occurrences of the same scalar, it would be cumbersome to do this for every query.

**Reply by aexol** [reply]
Yes I plan to move it to Chain do you don't have specify it every time

**Reply by aexol** [reply]
But moving to generation is not a good idea not typesafe at all

**Reply by ValentinH** [reply]
It actually depends on the use case. In my original issue, I'm talking about the `timestamp` type from Hasura which can always be set to `string` in TS.

**Reply by aexol** [reply]
Yeah but it is better to provide it to the function as a parameter to `Thunder` or `Chain`. Right now it is just in Chain execution result which is not so handy. But still, the PoC works that makes me happy :)

**Comment by ItsWendell** [comment]
What is the recommended way to create a custom global client? I did this for Hasura to support typescript with custom scalars:

```ts
import {
  Chain,
  GenericOperation,
  GraphQLTypes,
  InputType,
  OperationOptions,
  ScalarDefinition,
  ValueTypes,
} from "./generated/zeus";
import { Ops } from "./generated/zeus/const";

export * from "./generated/zeus";

const scalars = {
  jsonb: {
    encode: (e: unknown) => JSON.stringify(e) as string,
    decode: (e: unknown) => JSON.parse(e as string) as string,
  },
  JSON: {
    encode: (e: unknown) => JSON.stringify(e) as string,
    decode: (e: unknown) => JSON.parse(e as string) as string,
  },
  timestamptz: {
    encode: (e: unknown) => JSON.stringify(e) as string,
    decode: (e: unknown) => JSON.parse(e as string) as string,
  },
};

export const ZeusClient =
  (endpoint: string, secret: string) =>
  <
    O extends keyof typeof Ops,
    R extends keyof ValueTypes = GenericOperation<O>
  >(
    operation: O
  ) => {
    const chain = Chain(endpoint, {
      headers: {
        "x-hasura-admin-secret": secret,
      },
    })(operation);

    return <Z extends ValueTypes[R], SCLR extends ScalarDefinition>(
      o: Z | ValueTypes[R],
      ops?: OperationOptions & {
        scalars?: SCLR;
      }
    ): Promise<InputType<GraphQLTypes[R], Z, SCLR & typeof scalars>> =>
      chain(o as any, {
        ...ops,
        scalars: scalars,
      }) as any;
  };

export type ZeusClient = ReturnType<typeof ZeusClient>;

export default ZeusClient;
```

Seems a bit tedious, I was wondering if there is an easier way to accomplish this?

**Comment by aexol** [comment]
Of course. As I responded to the other issue. I will move scalars definition one level up as a Chain/Thunder param to make construction of those easier

**Comment by aexol** [comment]
Check graphql-zeus@5.0.4
https://zeus.graphqleditor.com/page/graphql/scalars.html

**Reply by aexol** [reply]
I moved scalars level up

**Comment by aexol** [comment]
forgot to publish - publishing now
